syscall/syscall.c

```diff
@@ -33,851 +34,1043 @@ typedef int (*ioctl_proc)(
+
+    return oe_sendmsg(sockfd, (struct oe_msghdr*)buf, flags);
+}
+
+OE_DEFINE_SYSCALL5(SYS_setsockopt)
+{
+    oe_errno = 0;
+    int sockfd = (int)arg1;
+    int level = (int)arg2;
+    int optname = (int)arg3;
+    void* optval = (void*)arg4;
+    oe_socklen_t optlen = (oe_socklen_t)arg5;
+    return oe_setsockopt(sockfd, level, optname, optval, optlen);
+}
+
+OE_DEFINE_SYSCALL2(SYS_shutdown)
+{
+    oe_errno = 0;
+    int sockfd = (int)arg1;
+    int how = (int)arg2;
+    return oe_shutdown(sockfd, how);
+}
+
+OE_DEFINE_SYSCALL3(SYS_socket)
+{
+    oe_errno = 0;
+    int domain = (int)arg1;
+    int type = (int)arg2;
+    int protocol = (int)arg3;
+    return oe_socket(domain, type, protocol);
+}
+
+OE_DEFINE_SYSCALL4(SYS_socketpair)
+{
+    oe_errno = 0;
+    int domain = (int)arg1;
+    int type = (int)arg2;
+    int protocol = (int)arg3;
+    int* sv = (int*)arg4;
+
+    return oe_socketpair(domain, type, protocol, sv);
+}
+
+#if defined(OE_SYS_stat)
+OE_DEFINE_SYSCALL2(SYS_stat)
+{
+    oe_errno = 0;
+    const char* pathname = (const char*)arg1;
+    struct oe_stat_t* buf = (struct oe_stat_t*)arg2;
+    return oe_stat(pathname, buf);
+}
+#endif
+
+OE_DEFINE_SYSCALL2(SYS_truncate)
+{
+    oe_errno = 0;
+    const char* path = (const char*)arg1;
+    ssize_t length = (ssize_t)arg2;
+
+    return oe_truncate(path, length);
+}
+
+OE_DEFINE_SYSCALL3(SYS_write)
+{
+    oe_errno = 0;
+    int fd = (int)arg1;
+    const void* buf = (void*)arg2;
+    size_t count = (size_t)arg3;
+
+    return oe_write(fd, buf, count);
+}
+
+OE_DEFINE_SYSCALL3(SYS_writev)
+{
+    oe_errno = 0;
+    int fd = (int)arg1;
+    const struct oe_iovec* iov = (const struct oe_iovec*)arg2;
+    int iovcnt = (int)arg3;
+
+    return oe_writev(fd, iov, iovcnt);
+}
+
+OE_DEFINE_SYSCALL1(SYS_uname)
+{
+    oe_errno = 0;
+    struct oe_utsname* buf = (struct oe_utsname*)arg1;
+    return oe_uname(buf);
+}
+
+#if defined(OE_SYS_unlink)
+OE_DEFINE_SYSCALL1(SYS_unlink)
+{
+    oe_errno = 0;
+    const char* pathname = (const char*)arg1;
+
+    return oe_unlink(pathname);
+}
+#endif
+
+OE_DEFINE_SYSCALL3(SYS_unlinkat)
+{
+    oe_errno = 0;
+    long ret = -1;
+    int dirfd = (int)arg1;
+    const char* pathname = (const char*)arg2;
+    int flags = (int)arg3;
+
+    if (dirfd != OE_AT_FDCWD)
+    {
+        oe_errno = OE_EBADF;
+        goto done;
     }
 
-    /* Unreachable */
+    if (flags != OE_AT_REMOVEDIR && flags != 0)
+    {
+        oe_errno = OE_EINVAL;
+        goto done;
+    }
+
+    if (flags == OE_AT_REMOVEDIR)
+        ret = oe_rmdir(pathname);
+    else
+        ret = oe_unlink(pathname);
 done:
     return ret;
 }
 
+OE_DEFINE_SYSCALL2(SYS_umount2)
+{
+    oe_errno = 0;
+    const char* target = (const char*)arg1;
+    int flags = (int)arg2;
+
+    (void)flags;
+
+    return oe_umount(target);
+}
+
+static long _syscall(
+    long num,
+    long arg1,
+    long arg2,
+    long arg3,
+    long arg4,
+    long arg5,
+    long arg6)
+{
+    oe_errno = 0;
```

Comment:
It seems now every syscall wrapper sets `oe_errno` to 0. I assume it's because those wrappers will be invoked directly in the following PRs and `_syscall` (and hence this line) can be bypassed. For the consistency, should be just remove this line?

Reply:
Removed. Added the comment explaning that the syscall implementation functions could be called directly, and hence they must each set oe_errno correctly.